### ISSUE TITLE
Runtime exception during initialization of SparkML model (One falsenode is pointing either to itself, either to another tree.)

### ISSUE THREAD

**Issue by harishankar-gopalan** [microsoft/onnxruntime#8008]
**Describe the bug**
Issue while loading a RandomForestClassifier model, exported as ONNX from PySpark using ONNX ml tools.

Exception trace seen:

> 2021-06-09 17:17:54.598724213 [E:onnxruntime:, inference_session.cc:1340 operator()] 
Exception during initialization: /onnxruntime_src/onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h:167 onnxruntime::ml::detail::TreeEnsembleCommon<ITYPE, OTYPE>::TreeEnsembleCommon(int, int, const string&, const std::vector<IT>&, int64_t, const std::vector<long int>&, const std::vector<long int>&, const std::vector<IT>&, const std::vector<long int>&, const std::vector<std::basic_string<char> >&, const std::vector<long int>&, const std::vector<long int>&, const std::vector<long int>&, const std::vector<IT>&, const string&, const std::vector<long int>&, const std::vector<long int>&, const std::vector<long int>&, const std::vector<IT>&) [with ITYPE = float; OTYPE = float; std::string = std::basic_string<char>; int64_t = long int] **One falsenode is pointing either to itself, either to another tree**.

**System information**
- OS Platform and Distribution: Linux 20.04
- ONNX Runtime installed from (source or binary): PyPI
- ONNX Runtime version: 1.8.0
- ONNX ML Tools version: 1.7.0
- Python version: 3.8
- PySpark version: 3.1.2
- Visual Studio version (if applicable): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A (using CPU inference)
- GPU model and memory: N/A

**To Reproduce**
- To reproduce the issue use the below python snippet to load the attached model  
   ```python
   def infer_from_onnx():
       input_list = [[randint(0, 20) for _ in range(32)]]
       sess = InferenceSession("<path to onnx model here>")
       print([x.name for x in sess.get_inputs()])
       input_name = sess.get_inputs()[0].name
       output_name = sess.get_outputs()[0].name
       pred_onx = sess.run([output_name], {input_name: numpy.array(input_list, numpy.float32)})
       print(pred_onx)
   ```
- Summary of the attached model as reported from PySpark (5 trees, 32 features)
  ```
  RandomForestClassificationModel: uid=RandomForestClassifier_b21fc18d5170, numTrees=5, numClasses=2, numFeatures=32
    DecisionTreeClassificationModel: uid=dtc_d5dde2bcb50e, depth=2, numNodes=5, numClasses=2, numFeatures=32
    DecisionTreeClassificationModel: uid=dtc_6a76eaab2e06, depth=2, numNodes=5, numClasses=2,numFeatures=32
    DecisionTreeClassificationModel: uid=dtc_1b65c6c01b09, depth=2, numNodes=5, numClasses=2, numFeatures=32
    DecisionTreeClassificationModel: uid=dtc_a6a141e31c41, depth=1, numNodes=3, numClasses=2, numFeatures=32
    DecisionTreeClassificationModel: uid=dtc_6d6a063bbb17, depth=1, numNodes=3, numClasses=2, numFeatures=32
  ``` 
- Additional details:
  - The issue does not occur when we generate a model of (2 trees and 32 features)
  - Performing inference using a model with 2 trees and 32 features works fine.
  - Tried quite a few combinations of number of trees and features, there doesnt seem to be any specific pattern. It looks like some combination of tree count and feature count throws this issue.

**Expected behavior**
With the exported ONNX model, the inference should be possible.

**Additional context**
The issue seems to have been discussed in this [issue](https://github.com/microsoft/onnxruntime/issues/4486) with no conclusive solution due to lack of model files. I have attached the model files and the code for reproducing it. 

Code for creating a PySpark model:
```python
spark = SparkSession.builder.getOrCreate()
def create_model():
    df = spark.read.csv("<path to attached data csv here>", header=True, inferSchema=True)

    vec_assembler = VectorAssembler(
        inputCols=["c"+str(i) for i in range(FEATURE_LEN)], outputCol="features"
    )

    data = vec_assembler.transform(df)

    rf = RandomForestClassifier(labelCol="label", featuresCol="features", numTrees=5)

    model: RandomForestClassificationModel = rf.fit(dataset=data)
    model.save("./dummy_spark_model/model/")
```

Code for exporting the model to onnx:
```python
def export_as_onnx():
    model = load_model()
    model_onnx = convert_sparkml(
        model,
        "Phish Classifier",
        [("features", FloatTensorType([1, FEATURE_LEN]))],
        spark_session=spark,
    )
    with open("./dummy_spark_model/model/sp_model.onnx", "wb") as f:
        f.write(model_onnx.SerializeToString())
```

Code for performing inference is already listed in the section [**To Reproduce**]

[onnxmodel.zip](https://github.com/microsoft/onnxruntime/files/6623577/onnxmodel.zip)
[features_32.csv](https://github.com/microsoft/onnxruntime/files/6628283/features_32.csv)

**Comment by yuslepukhin** [comment]
Cc: @xadupre

**Comment by xadupre** [comment]
The error message `One falsenode is pointing either to itself, either to another tree.` means the conversion went wrong, one tree is pointing to another one when dealing with missing value. This is an issue for onnxmltools. The bug fix may some time as we did not update sparkml converters for more than a year.

**Comment by harishankar-gopalan** [comment]
> The error message `One falsenode is pointing either to itself, either to another tree.` means the conversion went wrong, one tree is pointing to another one when dealing with missing value. This is an issue for onnxmltools. The bug fix may some time as we did not update sparkml converters for more than a year and the conversion first converts the tree into scikit-learn structure before using sklearn-onnx to convert it. I would suggest to use older version of sklearn-onnx to fix it in the mean time.

@xadupre Appreciate your prompt response. Is there any specific version of `onnxmltools` that you would recommend us to downgrade to ? Also could you elaborate the below a bit so that it will facilitate us to check if we can make any tweaks on the model side ?
`means the conversion went wrong, one tree is pointing to another one when dealing with missing value`

**Comment by xadupre** [comment]
I tried to replicate the bug here: https://github.com/onnx/onnxmltools/pull/471/files#diff-6339cc5d8aaddc264b67744d2ba358897fb08e7a80580e0e4bbecb9a5a68a704 without any success. Could you check I did not forget anything while replicating your example?

**Comment by harishankar-gopalan** [comment]
Hi @xadupre I went through the code and those seem to be equivalent to whatever I am doing, but I am able to reproduce the issue consistently. I have tried today multiple times with new models of same configuration and also switched the model loading from file to inline using `SerializeToString` method during inference (that was the only notable difference between your version of the code and mine). I am still facing the same error.
Also would like to know whether the model file that I had shared, did not throw any errors during inference ?

**Comment by xadupre** [comment]
Here is the model I got. Which version of sklearn-onnx (skl2onnx) are you using?

[model.zip](https://github.com/microsoft/onnxruntime/files/6688192/model.zip)

**Comment by harishankar-gopalan** [comment]
> Here is the model I got. Which version of sklearn-onnx (skl2onnx) are you using?
> 
> [model.zip](https://github.com/microsoft/onnxruntime/files/6688192/model.zip)

Hi @xadupre  thanks for the details. I am using *skl2onnx* version *1.8.0*. I should be able to change this to any specific version you can point me to.
Also FYI the attached *model.onnx* file does work fine for inference in my setup too. It might help investigate further if you could share the dependency versions you are using to perform the export to onnx, preferably the result `pip freeze` or `conda list` or that of any specific dependency you deem fit to be looked into ?

**Comment by R1ckyG** [comment]
Commenting on this to reopen. I'm also encountering the same issue with trying to use a randomforestclassifier converted from spark ml. I've noticed that it works when the data's label all belongs to a single class. Is there a work around for this issue? Let me know what files info would be useful for debugging.

**Comment by haiwi-song** [comment]
VectorAssembler + GBTClassifier + Pipeline，the same problem，no solution
